### ISSUE TITLE
mini batch training

### ISSUE THREAD

**Issue by luhaowang** [blei-lab/edward#320]
Hi,
Could you give some suggestions about how to do mini batch training for bayesian neural networks? I played with the n_minibatch parameters, using following code:
x_train = mnist.test.images
y_train = mnist.test.labels

x = ed.placeholder(tf.float32, [batch_size, IMAGE_PIXELS])
y = Normal(mu=bnns(x, W_0, W_1, W_2, b_0, b_1, b_2),
           sigma=0.1 * tf.ones([batch_size, NUM_CLASSES]))

data = {x:x_train, y:y_train}

**Comment by dustinvtran** [comment]
yes, that should work (syntactically). what is `num_classes`? is this for classification? what about 
```python
M =  # minibatch size
H =  # hidden layer size
D =  # number of features
K =  # number of class labels


W_0 = Normal(mu=tf.zeros([D, H]), sigma=tf.ones([D, H]))
W_1 = Normal(mu=tf.zeros([H, K]), sigma=tf.ones([H, K]))
b_0 = Normal(mu=tf.zeros(H), sigma=tf.ones(L))
b_1 = Normal(mu=tf.zeros(K), sigma=tf.ones(K))

def neural_network(x):
    h = tf.nn.tanh(tf.matmul(x, W_0) + b_0)
    h = tf.matmul(h, W_1) + b_1
    return tf.reshape(h, [-1])

x = ed.placeholder(tf.float32, [M, D])
y = Categorical(logits=neural_network(x))
``` 
during inference, you can do something like
```python
y_ph = ed.placeholder(tf.float32, [M])
inference = ed.KLqp({...}, data={y: y_ph})
...
for _ in range(inference.n_iter):
  inference.update(feed_dict={x: ..., y_ph: ....})
```

**Comment by luhaowang** [comment]
Hi Dustin, 
  Thank you for your reply. What I am doing is MNIST multi classification task. The output is one-hot type, For Categorical() function, I couldn't find nay module in Edward. You are tensorflow one?

**Comment by luhaowang** [comment]
I got an error in Multinomial()
Tensor conversion requested dtype int32 for Tensor with dtype float32: 'Tensor("Multinomial/p/Identity:0", shape=(10, 10), dtype=float32)'

Why should tensorflow convert dtype for multinomial? Did you see anything wrong with my code?

I posted my code as followed:

def bnns(x, W_0, W_1, b_0, b_1):
    h = tf.nn.relu(tf.matmul(x, W_0) + b_0)
    h = tf.nn.relu(tf.matmul(h, W_1) + b_1)
    return h

x = ed.placeholder(tf.float32, [batch_size, IMAGE_PIXELS])
y = Multinomial(n=batch_size,p=bnns(x, W_0, W_1, b_0, b_1))
y_ph = ed.placeholder(tf.float32, [batch_size])
data = {y:y_ph}

# INFERENCE
qW_0 = Normal(mu=tf.Variable(tf.random_normal([IMAGE_PIXELS, hidden1_units])),
              sigma=tf.nn.softplus(tf.Variable(tf.random_normal([IMAGE_PIXELS, hidden1_units]))))
qW_1 = Normal(mu=tf.Variable(tf.random_normal([hidden1_units, NUM_CLASSES])),
              sigma=tf.nn.softplus(tf.Variable(tf.random_normal([hidden1_units, NUM_CLASSES]))))
qb_0 = Normal(mu=tf.Variable(tf.random_normal([hidden1_units])),
              sigma=tf.nn.softplus(tf.Variable(tf.random_normal([hidden1_units]))))
qb_1 = Normal(mu=tf.Variable(tf.random_normal([NUM_CLASSES])),
              sigma=tf.nn.softplus(tf.Variable(tf.random_normal([NUM_CLASSES]))))

# CRITICISM
inference = ed.MFVI({W_0: qW_0, b_0: qb_0,
                     W_1: qW_1, b_1: qb_1},data)

**Comment by dustinvtran** [comment]
can you post the full script here? also, you can wrap your code with \`\`\` to correctly format the code snippets.
